### ISSUE TITLE
inherited NamedTuple with a generic function attribute requires alias

### ISSUE THREAD

**Issue by meshantz** [microsoft/pyright#3132]
**Describe the bug**
The following 3 `NamedTuple`s should be equivalent, but the third one results in an error.

```python
from typing import Callable, Coroutine, NamedTuple, TypeVar

T = TypeVar("T", Callable, Coroutine)
CT = Callable[[T], T]

thing1 = NamedTuple("thing1", [("f1", Callable[[T], T])])

class thing2(NamedTuple):
    f1: CT

class thing3(NamedTuple):
    f1: Callable[[T], T]
```

Error produced:
> Type variable "T" has no meaning in this context

**To Reproduce**
Code as above

**Expected behavior**
`thing3` should be interpreted the same way as `thing1` or `thing2`.

**VS Code extension or command-line**
VS Code extension: Pylance v2022.2.4

**Comment by erictraut** [comment]
A TypeVar must have a well-defined scope. PEP 484 documents two valid scopes for type variables and alludes to a third. The two documented scopes are classes and functions (`def` statements). The third is a generic type alias, which it doesn't specifically call out as a valid scope for a type variable, but all Python type checkers recognize it as such. You are attempting to use a type variable in a manner where none of those three scopes is present, so pyright is correct to flag it as an error.

The reason `thing2` works is because you have used a generic type alias `CT` without specifying a type argument, so that type argument is assumed to be `Unknown`. That's probably not what you intended. If you replace this with `CT[T]`, you will receive the same error that '"T" has no  meaning in this context'.

Arguably, `thing1` should also result in an error because you are using `T` in a manner where there is no defined scope.

`NamedTuple` classes don't support generics currently, so all entries within a `NamedTuple` must be concrete. Here's a way you can achieve what (I think) you're trying to do here:

```python
class thing4(NamedTuple):
    f1: CT[Callable] | CT[Coroutine]
```

**Comment by meshantz** [comment]
That's... frustrating.

Okay, I think I follow that. And from https://www.python.org/dev/peps/pep-0484/#scoping-rules-for-type-variables
this example seems very close to identifying what I'm violating:

```python
T = TypeVar('T')
S = TypeVar('S')

def a_fun(x: T) -> None:
    # this is OK
    y = []  # type: List[T]
    # but below is an error!
    y = []  # type: List[S]

class Bar(Generic[T]):
    # this is also an error
    an_attr = []  # type: List[S]

    def do_something(x: S) -> S:  # this is OK though
```

I believe your suggestion is correct. I am looking to ensure that the callable assigned to the `f1` attribute returns the same type as is passed to it.

Thanks for all your help here!
